mawabot/cogs/text/reddit.py

```diff
@@ -0,0 +1,185 @@
+
+        if self.token is None:
+            await self.refresh_token()
+
+        async with self.session.get(url, headers=self._headers()) as req:
+            if req.status == 401:
+                logger.debug('Reddit token out of date, refreshing...')
+                await self.refresh_token()
+
+                # Try again with new token
+                async with self.session.get(url, headers=self._headers()) as req:
+                    req.raise_for_status()
+                    data = await req.json()
+            else:
+                req.raise_for_status()
+                data = await req.json()
+
+        return data
+
+    async def refresh_token(self,):
+        logger.info('Getting new Reddit token')
+        reddit = self.bot.config['reddit']
+        async with self.session.post('https://www.reddit.com/api/v1/access_token',
+                                    auth=aiohttp.BasicAuth(reddit['app-id'], reddit['app-secret']),
+                                    data={'grant_type': 'client_credentials'}) as req:
+            req.raise_for_status()
+            data = await req.json()
+            self.token = data['access_token']
+
+    def get_image(self, item, channel):
+        # Check if nsfw images is given and channel is a nsfw channel
+        nsfw = item['over_18']
+
+        if not nsfw:
+            resolutions = item['preview']['images'][0]['resolutions']
+            image = resolutions[1]
+            image['url'] = image['url'].replace('&amp;', '&')
+        else:
+            if isinstance(channel, discord.abc.PrivateChannel) or channel.is_nsfw():
+                resolutions = item['preview']['images'][0]['resolutions']
+                image = resolutions[1]
+                image['url'] = image['url'].replace('&amp;', '&')
+            else:
+                return None
+        
+        embed = discord.Embed()
+        embed.title = item['title']
+        embed.url = 'https://www.reddit.com/' + item['permalink']
+        embed.set_image(url=image['url'])
+        embed.image.width = image['width']
+        embed.image.height = image['height']
+
+        return embed
+    
+    @commands.command()
+    @check_reddit
+    async def headpat(self, ctx):
+        items = await self.request('/r/headpats/random')
+        item = items[0]['data']['children'][0]['data']
+        
+        # If the image is nsfw and the channel isn't set as nsfw image is none
+        image = self.get_image(item, ctx.channel)
+
+        if image is not None:
+            await ctx.send(embed=image)
+        else:
+            # Retry once more incase it was just that one
+
+            items = await self.request('/r/headpats/random')
+            item = items[0]['data']['children'][0]['data']
+        
+            image = self.get_image(item, ctx.channel)
+
+            if image is not None:
+                await ctx.send(embed=image)
+            else:
+                await ctx.message.add_reaction('🔞')
```

Comment:
The NSFW checking is duplicate code. There should be an `async` helper method that posts or reacts similar to `get_image()`. Perhaps named `post_image()`.